File: frontend/src/api/routines.ts
/**
 * RoutineTask management client — Background Tasks page + Inbox.
 *
 * Mirrors `backend/app/schemas/agentive/routines.py` and:
 *
 *   GET    /api/agentive/routines
 *   GET    /api/agentive/routines/:id
 *   GET    /api/agentive/routines/:id/activity
 *   PATCH  /api/agentive/routines/:id
 *   POST   /api/agentive/routines/:id/cancel
 *   DELETE /api/agentive/routines/:id
 */

import api from './client';

export type RoutineStatus = 'active' | 'paused' | 'completed' | 'cancelled';
export type RoutineUpdateStatus = 'active' | 'paused';
export type LastRunStatus = 'success' | 'error' | 'skipped' | 'stopped';

export interface RoutineResponse {
  id: string;
  instruction: string;
  cron: string;
  timezone: string;
  status: RoutineStatus | string;
  write_scope: Array<
    { resource_type?: string; resource_id?: string } & Record<string, string>
  >;
  next_run_at: string | null;
  last_run_at: string | null;
  last_run_status: LastRunStatus | string | null;
  last_run_error: string | null;
  consecutive_failures: number;
  max_runs: number | null;
  run_count: number;
  thread_id: string;
  workspace_id: string;
  created_at: string | null;
  updated_at: string | null;
  /** True when this process has an in-flight origin=routine_task turn. */
  running?: boolean;
}

export interface RoutineListResponse {
  routines: RoutineResponse[];
  total: number;
}

export interface RoutineUpdateRequest {
  status?: RoutineUpdateStatus;
  instruction?: string;
  cron?: string;
  timezone?: string;
  max_runs?: number | null;
  clear_max_runs?: boolean;
}

export interface RoutineCancelResponse {
  id: string;
  status: 'cancelled';
}

export interface RoutineDeleteResponse {
  id: string;
  deleted: true;
}

export interface RoutineActivityLink {
  kind: string;
  id: string;
  label: string;
  href: string;
}

export interface RoutineActivityEvent {
  id: string;
  kind: string;
  ts: string | null;
  action: string | null;
  status: string | null;
  summary: string;
  reason: string | null;
  message_id: string | null;
  role: string | null;
  links: RoutineActivityLink[];
}

export interface RoutineActivityResponse {
  routine_id: string;
  thread_id: string;
  events: RoutineActivityEvent[];
  write_scope_links: RoutineActivityLink[];
}

export interface ListRoutinesParams {
  status?: RoutineStatus;
}

export async function listRoutines(
  params: ListRoutinesParams = {},
): Promise<RoutineListResponse> {
  const res = await api.get('/agentive/routines', { params });
  return res.data as RoutineListResponse;
}

export async function getRoutine(id: string): Promise<RoutineResponse> {
  const res = await api.get(`/agentive/routines/${encodeURIComponent(id)}`);
  return res.data as RoutineResponse;
}

export async function getRoutineActivity(
  id: string,
  limit = 40,
): Promise<RoutineActivityResponse> {
  const res = await api.get(
    `/agentive/routines/${encodeURIComponent(id)}/activity`,
    { params: { limit } },
  );
  return res.data as RoutineActivityResponse;
}

export async function updateRoutine(
  id: string,
  body: RoutineUpdateRequest,
): Promise<RoutineResponse> {
  const res = await api.patch(
    `/agentive/routines/${encodeURIComponent(id)}`,
    body,
  );
  return res.data as RoutineResponse;
}

export async function cancelRoutine(
  id: string,
): Promise<RoutineCancelResponse> {
  const res = await api.post(
    `/agentive/routines/${encodeURIComponent(id)}/cancel`,
  );
  return res.data as RoutineCancelResponse;
}

export async function deleteRoutine(
  id: string,
): Promise<RoutineDeleteResponse> {
  const res = await api.delete(
    `/agentive/routines/${encodeURIComponent(id)}`,
  );
  return res.data as RoutineDeleteResponse;
}

File: backend/app/schemas/agentive/routines.py
"""Schemas for agentive/api/routines.py — RoutineTask management REST surface.

Thin wire boundary over ``app.agentive.services.routine_tasks``. ``write_scope``
is readable on responses but intentionally omitted from ``RoutineUpdateRequest``
so the management UI cannot widen the grant without going through staging bless.
"""

from __future__ import annotations

from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field

RoutineUpdateStatus = Literal["active", "paused"]


class RoutineResponse(BaseModel):
    """Wire shape for a single RoutineTask row."""

    id: str
    instruction: str
    cron: str
    timezone: str
    status: str
    write_scope: List[Dict[str, str]] = Field(default_factory=list)
    next_run_at: Optional[str] = None
    last_run_at: Optional[str] = None
    last_run_status: Optional[str] = None
    last_run_error: Optional[str] = None
    consecutive_failures: int = 0
    max_runs: Optional[int] = None
    run_count: int = 0
    thread_id: str = ""
    workspace_id: str = ""
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    running: bool = False

    model_config = {"extra": "forbid"}


class RoutineListResponse(BaseModel):
    """List shape for GET /agentive/routines."""

    routines: List[RoutineResponse]
    total: int

    model_config = {"extra": "forbid"}


class RoutineUpdateRequest(BaseModel):
    """Body for PATCH /agentive/routines/{id}.

    ``write_scope`` is intentionally absent — display-only on the wire so
    widening cannot bypass staging bless.
    """

    status: Optional[RoutineUpdateStatus] = None
    instruction: Optional[str] = None
    cron: Optional[str] = None
    timezone: Optional[str] = None
    max_runs: Optional[int] = None
    clear_max_runs: bool = False

    model_config = {"extra": "forbid"}


class RoutineCancelResponse(BaseModel):
    """Wire shape for POST /agentive/routines/{id}/cancel."""

    id: str
    status: Literal["cancelled"] = "cancelled"

    model_config = {"extra": "forbid"}


class RoutineDeleteResponse(BaseModel):
    """Wire shape for DELETE /agentive/routines/{id}."""

    id: str
    deleted: Literal[True] = True

    model_config = {"extra": "forbid"}


class RoutineActivityLink(BaseModel):
    """Navigable resource mentioned by a routine run."""

    kind: str  # track | entry | app | other
    id: str
    label: str
    href: str

    model_config = {"extra": "forbid"}


class RoutineActivityEvent(BaseModel):
    """A scheduler change-event or chat turn attributed to this routine."""

    id: str
    kind: str  # run_event | message
    ts: Optional[str] = None
    action: Optional[str] = None
    status: Optional[str] = None
    summary: str
    reason: Optional[str] = None
    message_id: Optional[str] = None
    role: Optional[str] = None
    links: List[RoutineActivityLink] = Field(default_factory=list)

    model_config = {"extra": "forbid"}


class RoutineActivityResponse(BaseModel):
    """Activity feed for GET /agentive/routines/{id}/activity."""

    routine_id: str
    thread_id: str
    events: List[RoutineActivityEvent]
    write_scope_links: List[RoutineActivityLink] = Field(default_factory=list)

    model_config = {"extra": "forbid"}
